### PR TITLE
d&i four categorical operations

### PR DESCRIPTION
**Pull request by kamalsaleh**
* MereExistenceOfUniqueSolutionOfLinearSystemInAbCategory(a,b,c)
* MereExistenceOfUniqueSolutionOfHomogeneousLinearSystemInAbCategory(a,b)
* BasisOfSolutionsOfHomogeneousLinearSystemInLinearCategory(a,b)
* BasisOfSolutionsOfHomogeneousDoubleLinearSystemInLinearCategory(a,b,c,d)

and adding some derivations to them.

### CODE REVIEW

**Comment by mohamed-barakat** [comment]
Excellent!

**Comment by kamalsaleh** [comment]
The special case we talked about still needs some thought.

**Comment by mohamed-barakat** [comment]
After some consideration, I suggest the following renaming in a subsequent PR:

`MereExistenceOfUniqueSolutionOfLinearSystemInAbCategory ` ->` IsLinearSystemInAbCategoryUniquelySolvable`

`MereExistenceOfUniqueSolutionOfHomogeneousLinearSystemInAbCategory` -> `IsHomogeneousLinearSystemInAbCategoryUniquelySolvable`

**Comment by mohamed-barakat** [comment]
I know the renaming can violate the analogy to the naming convention in CAP. We can discuss this.

File: gap/Tools.gd [hunk 1]
```diff
@@ -94,6 +94,128 @@ DeclareOperation( "AddSumOfImagesOfAllMorphisms",
...
+
+DeclareOperation( "AddMereExistenceOfUniqueSolutionOfHomogeneousLinearSystemInAbCategory",
+                  [ IsCapCategory, IsList ] );
+
+##
+DeclareOperation( "BasisOfSolutionsOfHomogeneousLinearSystemInLinearCategoryOp",
+                   [ IsList, IsList, IsCapCategory ] );
+
+#! @Description
+#! The arguments are two lists $\alpha$ and $\beta$ of morphisms in some linear category over commutative ring.
```
Comment (mohamed-barakat) [inline gap/Tools.gd:157]: two lists -> two lists of lists

File: gap/Tools.gd [hunk 2]
```diff
@@ -94,6 +94,128 @@ DeclareOperation( "AddSumOfImagesOfAllMorphisms",
...
+
+DeclareOperation( "AddBasisOfSolutionsOfHomogeneousLinearSystemInLinearCategory",
+                  [ IsCapCategory, IsList ] );
+
+##
+DeclareOperation( "BasisOfSolutionsOfHomogeneousDoubleLinearSystemInLinearCategoryOp",
+                   [ IsList, IsList, IsList, IsList, IsCapCategory ] );
+
+#! @Description
+#! The arguments are four lists $\alpha$, $\beta$, $\gamma$, $\delta$ of morphisms in some linear category over commutative ring.
```
Comment (mohamed-barakat) [inline gap/Tools.gd:190]: four lists -> four lists of lists

File: gap/Tools.gd [hunk 3]
```diff
@@ -94,6 +94,128 @@ DeclareOperation( "AddSumOfImagesOfAllMorphisms",
...
+DeclareOperation( "AddMereExistenceOfUniqueSolutionOfLinearSystemInAbCategory",
+                  [ IsCapCategory, IsList ] );
+
+##
+DeclareOperation( "MereExistenceOfUniqueSolutionOfHomogeneousLinearSystemInAbCategoryOp",
+                   [ IsList, IsList, IsCapCategory ] );
+
+#! @Description
+#! but the output is <C>true</C> if the homogeneous system has only the trivial solution,
+#! <C>false</C> otherwise.
```
Comment (mohamed-barakat) [inline gap/Tools.gd:134]: , and <C>false</C> otherwise

File: gap/Tools.gd [hunk 4]
```diff
@@ -94,6 +94,128 @@ DeclareOperation( "AddSumOfImagesOfAllMorphisms",
...
+###################################
+
+##
+DeclareOperation( "MereExistenceOfUniqueSolutionOfLinearSystemInAbCategoryOp",
+                   [ IsList, IsList, IsList, IsCapCategory ] );
+
+#! @Description
+#! Like <C>SolveLinearSystemInAbCategory</C>,
+#! but the output is simply <C>true</C> if a unique solution exists,
+#! <C>false</C> otherwise.
```
Comment (mohamed-barakat) [inline gap/Tools.gd:110]: , and false otherwise

**Pull request merged**